src/timestamp_generator.hpp

```diff
@@ -61,16 +61,24 @@ class ServerSideTimestampGenerator : public TimestampGenerator {
 
 class MonotonicTimestampGenerator : public TimestampGenerator {
 public:
-  MonotonicTimestampGenerator()
+  MonotonicTimestampGenerator(int64_t warning_threshold_us = 1000000,
+                              int64_t warning_interval_ms = 1000)
     : TimestampGenerator(MONOTONIC)
-    , last_(0) { }
+    , last_(0)
+    , last_warning_(0)
+    , warning_threshold_us_(warning_threshold_us)
+    , warning_interval_ms_(warning_interval_ms) { }
 
   virtual int64_t next();
 
 private:
   int64_t compute_next(int64_t last);
 
   Atomic<int64_t> last_;
+  Atomic<int64_t> last_warning_;
+
+  const int64_t warning_threshold_us_;
+  const int64_t warning_interval_ms_;
```

Comment:
> It shows an intent for valid argument values.

I used to think this too, but for the pitfall shown in the above example it's not worth it. If a user passes a negative number the value turns into a really large number and it's not possible to detect an error. It's better to used signed values and return an explicit error (or at least document and correct it).